### ISSUE TITLE
Optional NaN canonicalization mode?

### ISSUE THREAD

**Issue by NikVolf** [pepyakin/wasmi#19]

**Comment by Recmo** [comment]
I'm assuming this refers to the idea mentioned here as a solution to the non-determinism problem: https://github.com/WebAssembly/design/issues/582#issuecomment-191318866

> For NaN bits, it'd be straightforward for a preprocessor to inject NaN canonicalization code after every floating-point computation, and that would suffice.

Doing this after every operation might be excessive, NaNs with different bit patterns still mostly behave the same. It should be enough to canonicalize in the few places where they don't: `reinterpret`, `store`, etc. There is a list here: https://github.com/WebAssembly/design/blob/master/Rationale.md#nan-bit-pattern-nondeterminism

**Comment by NikVolf** [comment]
Yep, the motivation is to prevent non-determinism still having some base2 floats.

it's not very important atm, since we can ban floats on the embedder  and/or use  soft floats in the user space

But still might be useful!

**Comment by pepyakin** [comment]
> Although it would be fun to write a contract that pays out when someone mines it with ARM instead of x86. Might be good for an ARM node bounty! 🤡

Interesting idea, but how would you actually proof that? : D

**Comment by MaxGraey** [comment]
I don't think NaN is only non-determinism part of float points. operations like `sqrt` and `inverse sqrt (rsqrt)` implement differently even on x86 by some vendors. ARM for example flush all subnormal values to zero (FTZ and DAZ modes). Conversion from floats to integers also behave differently on hardware and by compiler's mode.

**Comment by sunfishcode** [comment]
Wasm's `sqrt` really is deterministic; while other languages can use x87 which rounds things like sqrt differently in some cases, wasm doesn't permit this, and this is [specifically tested for in the spec testsuite](https://github.com/WebAssembly/spec/blob/master/test/core/float_misc.wast#L524).

The ARM subnormal issue is only on 32-bit ARM, and only NEON, not VFP, so wasm's current scalar operations [aren't](https://github.com/WebAssembly/spec/blob/master/test/core/f32.wast) [affected](https://github.com/WebAssembly/spec/blob/master/test/core/f64.wast). This is a potential concern for [the SIMD proposal](https://github.com/WebAssembly/simd/blob/master/proposals/simd/SIMD.md), however.

WebAssembly also fully specifies the behavior of conversion from floats to integers, and the behavior in all the cases where hardware is known to differ is [covered in the testsuite too](https://github.com/WebAssembly/spec/blob/master/test/core/conversions.wast).

So while wasm could add rsqrt approximations or other things in the future, and SIMD will have to cope with 32-bit ARM NEON, in the current spec, NaN bit patterns really are the only floating-point nondeterminism.

**Comment by MaxGraey** [comment]
@sunfishcode Thank you for so detail clarifying! Now you had dispelled any doubts)

**Comment by Robbepop** [comment]
I thought about how this could be implemented efficiently for Wasmi.

## Design

1. Add `Config::canonicalize_nans(flag: bool)` to `wasmi::Config`.
    - This allows to enable or disable this feature on a per-`wasmi::Engine` basis.
2. Add `wasmi_ir::Instruction` variants for the following float instructions that require NaN canonicalization:
    - `{f32, f64}.sub`
    - `{f32, f64}.mul`
    - `{f32, f64}.div`
    - `{f32, f64}.sqrt`
    - `{f32, f64}.min`
    - `{f32, f64}.max`
    - `{f32, f64}.copysign`
    - `{f32, f64}.neg`
    - **Note:** NaN canonicalization needed only for intermediate computations:
        - `{i32, i64}.trunc_f{32,64}_{s,u}`
    - **Note:** No NaN canonicalization needed for:
        - `{f32,f64}.cmp` (e.g. `f32.eq`, `f64.lt`)
        - `{f32,f64}.branch_cmp` (fused instructions)
3. During Wasmi translation the new variants are emitted instead of the existing ones if NaN canonicalition is enabled.
4. The Wasmi executor performs NaN canonicalization for the new `Instruction` variants.

This way no performance is lost when NaN canonicalization is disabled but users who require the additional determinism can enable it with minor runtime costs since there won't be runtime checks during execution.

**Comment by Robbepop** [comment]
Design-change: unlike Wasmtime we should not add `Config::canonicalize_nans` but instead add a create feature `canonicalize_nans` to `wasmi` to finally support NaN canonicalization all while avoiding effects on binary artifact size and translation/execution performance.

Wasmtime being a JIT compiler can get away with a more flexible approach to solve this issue since its translation generates machine code whereas Wasmi's translation generates Wasmi IR for which we'd need to add new operators to support NaN canonicalization efficiently.

A similar design was chosen in the `tinywasm` interpreter.

**Comment by Robbepop** [comment]
Only took ~8 years to close this. :D
